Cs/utility associations PopupElement

changanxian opened the pull request.

jenmerritt commented on example/lib/main.dart:
Could you just explain why we've brought `GoRouter` in specifically for this popup view enhancement - are we requiring developers to have a certain type of routing package in order to be able to use the popup view now? Or is this a totally non-related change?

jenmerritt commented on example/lib/main.dart:
```diff
@@ -32,14 +34,34 @@ void main() {
   }
 
+  // Configure routes with GoRouter
+  final GoRouter _router = GoRouter(
```
Analyzer warning about the type annotation here, and local variable have `_`

jenmerritt commented on lib/src/popups/popup_view.dart:
I'm getting a crash if I keep navigating through maybe 4/5 levels deep.

```
══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
The following assertion was thrown while handling a gesture:
You have popped the last page off of the stack, there are no pages left to show
'package:go_router/src/delegate.dart':
Failed assertion: line 175 pos 7: 'currentConfiguration.isNotEmpty'

When the exception was thrown, this was the stack:
#2      GoRouterDelegate._debugAssertMatchListNotEmpty (package:go_router/src/delegate.dart:175:7)
#3      GoRouterDelegate._completeRouteMatch (package:go_router/src/delegate.dart:196:5)
#4      GoRouterDelegate._handlePopPageWithRouteMatch (package:go_router/src/delegate.dart:154:7)
#5      _CustomNavigatorState._handlePopPage (package:go_router/src/builder.dart:447:42)
#6      NavigatorState.pop (package:flutter/src/widgets/navigator.dart:5576:28)
#7      NavigatorState.popUntil (package:flutter/src/widgets/navigator.dart:5619:7)
#8      _isRouteInStack (package:arcgis_maps_toolkit/src/popups/popup_views/utility_association_result.dart:219:46)
#9      _navigateToAssociationPopupPage (package:arcgis_maps_toolkit/src/popups/popup_views/utility_association_result.dart:165:9)
#10     _UtilityAssociationResultState.build.<anonymous closure> (package:arcgis_maps_toolkit/src/popups/popup_views/utility_association_result.dart:59:20)
#11     _InkResponseState.handleTap (package:flutter/src/material/ink_well.dart:1204:21)
#12     GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:345:24)
#13     TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:758:11)
#14     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:383:5)
#15     BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:314:7)
#16     PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:721:9)
#17     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:97:12)
#18     PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:142:9)
#19     _LinkedHashMapMixin.forEach (dart:_compact_hash:764:13)
#20     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:140:18)
#21     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:130:7)
#22     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:528:19)
#23     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:498:22)
#24     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:473:11)
#25     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:437:7)
#26     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:394:5)
#27     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:341:7)
#28     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:308:9)
#29     _invoke1 (dart:ui/hooks.dart:346:13)
#30     PlatformDispatcher._dispatchPointerDataPacket (dart:ui/platform_dispatcher.dart:467:7)
#31     _dispatchPointerDataPacket (dart:ui/hooks.dart:281:31)
(elided 2 frames from class _AssertionError)

Handler: "onTap"
Recognizer:
  TapGestureRecognizer#894cb
```

changanxian replied:
This is partially caused by the popup example app if you don't modify it. The PopupView is shown in the bottom sheet in the example.  See my comments in the issue. We need to discuss this.

jenmerritt reviewed (commented):
This will take me a bit of time to review, need to get the other SDKs test apps running.

A few initial comments.

Also I will write up issues to cover these things separately:
- Verifying all styling fits in with the theme settings (I'm not going into detail in this PR)
- Review documentation (again not going into detail in this PR)

changanxian commented:
Changes since last review:

Page navigation has been implemented and integrated. The PopupView is still used as before; however, when the popup element is a utility popup element, a navigation stack is now established and used for navigation.

PaulAllanSturm commented on lib/src/popups/utility_associations_popup_element_view.dart:
```diff
@@ -0,0 +1,262 @@
+
+    _fetchAssociationsFuture = widget.popupElement
+        .fetchAssociationsFilterResults()
+        .timeout(
```
I don't think we need the timeout any more, now that we fixed the underlying problem with network requests getting stalled.

jenmerritt reviewed (approved):
Thanks @changanxian - not getting that exception anymore.

After testing, I think the functionality is there - I'm seeing the same data in terms of routing through as I do with the same data in the Swift toolkit. I think there's some refinement to do with the styling, as we haven't got consistency between the views as we go through (like the "x" and going back displays) - but we have got the ability to work through all the data and features. So I think the Navigator wrapper you've implemented is the right approach for this PR.

I've written up separate issues for:
- Refining the styling - and maybe we get Paul to take a look at this next sprint with fresh eyes
- Refining the documentation (so we don't have to worry about that in this PR)

changanxian commented:
@jenmerritt @PaulAllanSturm Thanks for your review and feedback.
